### ISSUE TITLE
Some ReadXX and WriteXX API on Marshal are not implemented

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/coreclr#10442]
The following API's on `Marshal` are not implemented and apparently should be. @Petermarcu found them during his experiments. 

```
 public static byte ReadByte([MarshalAs(UnmanagedType.AsAny), In] Object ptr, int ofs)
 public static int ReadInt32([MarshalAs(UnmanagedType.AsAny), In] Object ptr, int ofs)
 public static long ReadInt64([MarshalAs(UnmanagedType.AsAny), In] Object ptr, int ofs)
 public static short ReadInt16([MarshalAs(UnmanagedType.AsAny), In] Object ptr, int ofs)
 public static void WriteByte([MarshalAs(UnmanagedType.AsAny), In, Out] Object ptr, int ofs, byte val)
 public static void WriteInt16([MarshalAs(UnmanagedType.AsAny), In, Out] Object ptr, int ofs, short val)
 public static void WriteInt32([MarshalAs(UnmanagedType.AsAny), In, Out] Object ptr, int ofs, int val)
 public static void WriteInt64([MarshalAs(UnmanagedType.AsAny), In, Out] Object ptr, int ofs, long val)
```

**Comment by jkotas** [comment]
cc @yizhang82

**Comment by yizhang82** [comment]
@danmosemsft These are AsAny variants that are by design to throw PNSE. Is there any value in supporting them and do we have evidence of people using these particular overloads? People should be using the IntPtr versions.

**Comment by Petermarcu** [comment]
Is there a reason these cannot be supported? Why are they by design? I hit it while trying to get desktop assemblies running on .NET Core which is something we're trying to enable in 2.0. Existing packages from NuGet should just work. We only want to throw PNSE when we can't do it. After I hit it, I looked in the NuGet package database and found that other popular libraries were calling these methods.

**Comment by yizhang82** [comment]
@Petermarcu Can you share the Nuget package you are trying?

**Comment by Petermarcu** [comment]
I was using WinForms as a proxy. I think this was the most popular one that showed up in the report: https://www.nuget.org/packages/log4net/

In general though, we don't want people to hit PNSE exception unless it's obvious why something shouldn't/couldn't be supported on a platform. Is there something about these API's that makes them inherently unimplementable on .NET Core?

**Comment by Petermarcu** [comment]
I had a naive implementation of these that got me unblocked on my prototyping that got the IntPtr from for argument and just called the other methods. Jan told me that wasn't going to work in all cases.

**Comment by yizhang82** [comment]
@Petermarcu Right. There are quite a few cases that needs to be supported for Write/Read variants for AsAny. These APIs are badly designed and has horrible performance - every time it reads a byte it creates a copy of the object. There are also other interop features we have deprecated in .NET Core, such as IDispatch/Automation support, VARIANT/SafeArray marshaling, etc. If we found good usage evidence we can revisit those. You do get them more or less for free in Windows CoreCLR but for other runtime/platforms it is very expensive to get them implemented. Having said that, I think this is a good opportunity to reexamine the decisions we've made in the past and see if it still makes sense. We can also collect feedback and revisit this in next netstandard.

**Comment by jkotas** [comment]
FWIW, Mono throws PNSE for these APIs as well.

**Comment by yizhang82** [comment]
@jkotas Thanks. This is very good data point.
@Petermarcu Can you share call stack? I couldn't find Marshal.Read/Write calls in log4net.dll 2.0.8

**Comment by Petermarcu** [comment]
I believe this is in .NET Standard 2.0 right? I'm hoping to get .NET Core as close to complete for API's that are in .NET Standard 2.0 as possible with .NET Framework as the guiding light. We're looking at NuGet data as one data point but in general trying to make sure there is as little friction as possible for people picking up .NET Framework assets that already exist and using them on .NET Core.

How many of the features you mentioned are in types that we decided to put in .NET Standard 2.0?

My experiment was a couple of weeks ago now and I shared my notes with Dan.  I could probably get a repro if I can set aside some time so you could see the callstack. I feel like it was in some of the GDI+ interop code in system.drawing that I was seeing this.

**Comment by Petermarcu** [comment]
yeah, here is some example code from where it was getting hit: https://referencesource.microsoft.com/#System.Drawing/commonui/System/Drawing/Advanced/Font.cs,470

**Comment by danmosemsft** [comment]
Yes they are in NS since 1.1 in fact.
https://apisof.net/catalog/System.Runtime.InteropServices.Marshal.ReadInt16(Object,Int32)

**Comment by danmosemsft** [comment]
https://referencesource.microsoft.com/#System.Drawing/commonui/System/Drawing/Advanced/Font.cs,853

**Comment by Petermarcu** [comment]
Yeah, there is also a community project going on right now and we're considering adding System.Drawing to corefx. Not having these API's would make it harder to make System.Drawing work on .NET Core and that is an exact proxy for the pain we really don't want people to have when porting Full.NET Code that should work on .NET Standard 2.0 over to .NET Core which supports .NET Standard 2.0.

**Comment by Petermarcu** [comment]
I'd like to get these implemented on coreclr. If future platforms can't support them, that's where I want to throw PNSE. Any platform that can support .NET Standard, should do it the best it can and .NET Core is our flagship example of another .NET Platform implementing the standard.

**Comment by Petermarcu** [comment]
If we want to deprecate, we should do so in the standard.

**Comment by danmosemsft** [comment]
@yizhang82 are you open to implementing these (as far as is feasible)?

**Comment by yizhang82** [comment]
@danmosemsft I'm open to it. Let's find some time to chat about this. For the System.Drawing case, one option is to fix the System.Drawing code - and we'll also be making it faster in the process since the API always make a coyp of the struct before read/writing...

**Comment by danmosemsft** [comment]
Yep -- Peter's point of course is that that's not the only code out there that calls it, and we start dealing with 1000 paper cuts.

**Comment by Petermarcu** [comment]
Yeah, I'm more worried about making sure people just work and don't discover platform not supported exceptions when pulling in legacy code. Our goal to deprecate API's or get people on a better path is a fine goal but PNSE isn't the mechanism for that.

**Comment by jnm2** [comment]
Is this something that https://github.com/terrajobst/platform-compat should warn about? /cc @karelz

**Comment by danmosemsft** [comment]
@jnm2 yes and it does

@yizhang82 do you plan to assign to a developer so we can get this done for zbb...?

**Comment by yizhang82** [comment]
I'm still not yet convinced this should be done for netstandard 2.0. Especially that Mono doesn't support it today (throw PNSE). 

Implement this on top of CoreCLR is easy (since AsAny marshaling is there, though we do need to add tests and I'm not sure if we have them anywhere yet). However supporting this in .NET native would require additional ~2 week of work, and not to mention that it generates work for Mono as well. 

The System.Drawing usage seems to pretty horrifying by itself - why would someone want to define their own version of LOGFONT and pass it down as Object to ToLogFont? The doc for [ToLogFont](https://msdn.microsoft.com/en-us/library/2k3dsz1b(v=vs.110).aspx) even links to the native LOGFONT struct!

One option that @jkotas suggested is to support a scoped version (such as only blittable structs, string, etc) with more straight-forward semantics.

@danmosemsft Do you have a list of library that we care about that use these APIs? Once we have a bit more data I can see what a scoped version would look like and we can easily add it into CoreCLR, and other runtimes would also have a much easier time implementing this as well.

**Comment by danmosemsft** [comment]
@yizhang82 I have no such data. We only know of System.Drawing and that 0.3% of netfxcompat lab (whatever that data comes from) use some of them eg https://apisof.net/catalog/System.Runtime.InteropServices.Marshal.ReadInt16(Object,Int32)

**Comment by Petermarcu** [comment]
Lets get the easy work done before ZBB for coreclr. We can figure out what to do about .NET Native later.

**Comment by karelz** [comment]
If there are known incompatibilities (e.g. with Mono), please document them here: https://github.com/dotnet/corefx/wiki/ApiCompat

**Comment by yizhang82** [comment]
In principle I agree we should be compatible with desktop. However having a low usage and the fact Mono has the same behavior as CoreCLR today is probably good indication that this is a rather low priority API. Having said that, I don't mind getting this API into CoreCLR assuming that we are OK with accumulating more debt into .NET native and Mono down the road. I'll get it in this week.

**Comment by yizhang82** [comment]
Fixed by https://github.com/dotnet/coreclr/pull/11251
